lib/scope.go

```diff
@@ -91,20 +116,38 @@ func (s *scope) CollectionSet() collection.Set {
 	return s.inst.collectionSet
 }
 
+// ReplaceParentContext returns a copy of the scope bound to a new parent
+// context
+func (s *scope) ReplaceParentContext(newParent context.Context) scope {
```

Comment:
I've changed this to not use `context.Background`, and instead accept a new parent param. We should almost always be re-rooting these onto the application context, so workflow execution has a chance to clean up on process shutdown

Reply:
You are completely correct that this should be using `AppContext` instead of `context.Background`, and thanks for catching that bug in my original change. However, I don't think this should be `ReplaceParentContext(newParent)` and would rather it be `DetachParentContext`. Here's my reasoning:

For `DetachParentContext`, there's a very specific use-case it's covering: "this kicks off a job that runs in the parent, so my parent context's cancellation should not affect that job". For the new version `ReplaceParentContext`, I can't think of a case where some arbitrary context (aside from the AppContext) would be used to replace what is in the scope. It's too general of a tool, and so introduces a foot-gun. It makes it easy to reproduce my original error, by letting new code incorrectly call `ReplaceParentContext(context.Background())` when it should be using the `AppContext`. 

A general principle I like to follow, when there's code that looks like this:

```
obj.method(obj.thing(), other_params)
```

That is, a situation where `obj` has a `method` and some data that can be passed to that method. If it's the case that `obj.thing()` is the only valid thing to pass into `method`, and using some other value would be an error, it's preferable to just to eliminate that parameter, and have the implementation get the `thing()` in its body. In order words, do this:

```
obj.method(other_params)

...

func (o *object) method(params) {
  thing := o.thing()
  ...
}
```

This is exactly what you did with `sendEvent` in this same PR, by moving the `bus` into the implementation. There's no need to allow an arbitrary bus parameter when the one owned by the scope is the correct thing to use.

Now I could be mistaken, and we may want to have arbitrary contexts get passed to `ReplaceParentContext` at some point in the future. But (1) I can't think of that use-case off the top of my head, and (2) I'd rather avoid the potential errors until we observe evidence of that use-case.